### ISSUE TITLE
support for SXL modules

### ISSUE THREAD

**Issue by emiltin** [rsmp-nordic/rsmp_core#62]
Based on working group discussions:

**Background**
We currently have only one standardized SXL, the one of traffic lights. We would like to standardize more for eg. sensors, variable message signs, etc.
But we want to avoid duplicating specifications between various SXL. Instead we want to reuse specs.

**Idea**
Let's look at a more modular approach, so that equipment can implement what is relevant, and we avoid SXL duplication. Today the SXL for traffic lights include message for traffic data. A dedicated traffic counter today has no standardized SXL, and will typically end up copying from the traffic light specifications.

A better way would be to define modules. eg:

- traffic light controller
- traffic sensor
- variable message sign

A traffic light can then implement the TLC module, and optionally the traffic sensor module. A traffic counter can implement the traffic sensor module.

We also have variable messages signs that include a traffic counter, and uses it's own custom SXL with parts copied from the VMS projects and the trafffic light SXL. This could be improved by defining a VMS module.  

Extending this line of thinking, the current input/output messages in the traffic light SXL would fit in a general "input/output" module, which could be very useful on many other types of equipment. It could also be used on in its own for simple types of equipment where you can activate something and check it's status, eg. a pulley.

**Components**
A site can have many components, of different types. Each component type should be able to implement diffferent modules. For example the main component might a VMS and implement the vms module, while an integrated traffic counter would be a separate component, implementing the traffic sensor module.

**Handshake**
When a site connects, it should tell what modules it support, as part of the handshake sequence. This should be done in way that is backward-compatible, so if the site connects to a supervisor that still uses an older version of core, they can agree to use an older core version without issues.

If e.g. 3.2 or higher is agreed during the handshake, we could required that the site sends an list of component types and the modules they support. It should also be possible to also request this list later.

**Component and module types**
The site should provide info about:
1. component types and the modules they support
2. components and their types

**Modules**
- Input/Output: read inputs and activate outputs. Could be used for e.g. push buttons, pulleys, bridge sensors, etc.
- Traffic Light Controller: Everything specific to traffic lights, e.g. signal plans, cycle time, etc.
- Variable Message Sign: Writing in displays. Should suport things like predefined messages, text, images.
- Traffic sensor: traffic data collected by a sensor, or a groups of sensors

Thinking about what would be common between types of equipment, we might consider:
- Base: restart, clock, version, etc.
- Developer: Programmatically trigger alarms, get debug info, statistics, etc.

**Comment by emiltin** [comment]
This should be include in 3.3.0...?

**Comment by emiltin** [comment]
> Each component type should be able to implement different modules. 

I think it's clearer to say that modules defines types and messages, and messages target a specific type of component:

```mermaid
flowchart TD
  Parent[Parent Module]
  Module[Module]
  Type[Component Type]
  Message[Message]

  Module -->|extends| Parent
  Module -->|defines| Type
  Module -->|defines| Message
  Message -->|targets| Type
```

A submodule extends a parent module and can target all component types defines in the parent module.

**Comment by emiltin** [comment]
If we need modules to be able to target component types defined in other modules, we could use a dependency mechanism. E.g. If module A depends on module B, it can target component types defined in B.

I we need a module to be able to target any component types, we can use a wildcard mechanism, i.e. when a module defines a message, the target component type can be "any", or a list of component types. This could make sense for generic/system messages that would operate on any components.

I strongly believe we should use types that are clearly linked to the modules where they are defined, so that when a supervisor receives a component list, the type of a component is unambiguous.

Since modules are independent I think we should allow them to define component types with the same name. This is not a problem as long as the type includes the module name, e.g.:

```
vms/lamp
tlc/lamp
```

**Comment by emiltin** [comment]
The sub module mechanism means you can define a hierarchy or "family" of modules. Component types defined in parent modules can be used by all sub modules. 

Example:
You want to split a large module into smaller ones, but they all target the same component types.

To do this you define a parent module which defines the types.
Sub modules define message related to specific topics, and targeting the same component types, define in the parent:

```
/tlc   => defines types tc, dl, sg
/tlc/mode
/tlc/schedule
/tlc/programs
/tlc/prediction
```

**Comment by emiltin** [comment]
It needs to be backward compatible.

## Single Module
A device that supports only one module must still must include the "modules" list in it's Versions message:

```json
{
     "mType": "rSMsg",
     "type": "Version",
     "mId": "6f968141-4de5-42ff-8032-45f8093762c5",
     "RSMP": [
         { "vers": "3.1.1" },
         { "vers": "3.1.2" }
     ],
     "siteId": [
         { "sId": "O+14439=481WA001" }
     ],
     "SXL": "1.3.0",
     "modules": [
         { "id": "tlc", "version": "1.3.0" }
      ]
}
```

This ensure that the module name is communicated to the supervisor. Note that the module name is already specified in our SXL yaml source, e.g:
https://github.com/rsmp-nordic/rsmp_schema/blob/main/schemas/tlc/1.2.1/sxl.yaml#L3

A module defines component types e.g. the tlc module defines the types tc, sg, dl:
https://rsmp-nordic.github.io/rsmp_sxl_traffic_lights/1.2.1/sxl_traffic_light_controller.html#object-types

A module also defines the component types targeted by each message, e.g:
https://rsmp-nordic.github.io/rsmp_sxl_traffic_lights/1.2.1/sxl_traffic_light_controller.html#commands

## Format A component ids
The site can use component ids in the older format A, e.g. "KK+AG0503=001TC000". In the ComponentList:

The site still sends a ComponentList message, listing the components, but using format A for ids:

```json
[
   {"id": "KK+AG0503=001TC000", "type": "tlc/tc", "name": "Grand Plaza"},
   {"id": "KK+AG0503=001SG000", "type": "tlc/sg", "name": "North"},
   {"id": "KK+AG0503=001SG001", "type": "tlc/sg", "name": "South"},
   {"id": "KK+AG0503=001DL000", "type": "tlc/dl", "name": "Busses Northgoing"},
]
```

With format A, you cannot refer to a group of components. But you can filter by the type, e.g. "sg", either explicity or implicitely, eg. even though you target S0001 to the tc (main) component it gets data for all components of type sg. With format B, it still only gets data for components of type sg, but you can target a path like "/sg/north/" to get data for a subset of signal groups, or target "/tc" or "/" to get data for all signal groups.


## Mixing
Modules and format B can be used independently. You can use neither, one of them or both.

**Comment by emiltin** [comment]
Could we call it simply "modules" instead "sxl modules"? The name "SXL" has always been a bit of a tongue-twister..

**Comment by emiltin** [comment]
Currently, the meaning of bits in AggregatedStatus messages is defined in the SXL. My understanding is that only the main component sends aggregated status? Or does each grouped object send it? @otterdahl 

With modules, we need to be clear about which module(s) defines and sends aggregated status bits. Only once per site, or possibly from several components?

**Comment by otterdahl** [comment]
> Currently, the meaning of bits in AggregatedStatus messages is defined in the SXL. My understanding is that only the main component sends aggregated status? Or does each grouped object send it? @otterdahl 

Each grouped object sends it. This way, a single SXL can define multiple sites.

> With modules, we need to be clear about which module(s) defines and sends aggregated status bits. Only once per site, or possible from several components?

If we remove support for multiple sites (which I think we should), we can simply define that only the main component can send aggregated status, making this simpler.

**Comment by emiltin** [comment]
> If we remove support for multiple sites (which I think we should), we can simply define that only the main component can send aggregated status, making this simpler.

Right. This implies that sub components never send aggregated status, only the root (main) component?

**Comment by otterdahl** [comment]
> Right. This implies that sub components never send aggregated status, only the root (main) component?

Yes

**Comment by emiltin** [comment]
I started a PR to support SXL modules at #256.

**Comment by marcgarba** [comment]
To fuel reflection, regarding component types and their link to modules, instead of having to go through sub-module dependencies,
we could independently create family modules for component types, and then the signal modules would refer to them.

This seems more flexible and simpler, and above all less restrictive allowing us for example to manage cases where there are two equivalent and exclusive modules (which would use the same types of components).

I am trying this diagram (for the first time...).

```mermaid
flowchart TD
    A[Module SXL] -->|defines| B[Signals]
    B --> C[ComponentsTypes Family]
    A --> C

```
And at RSMP core messages level, we would have this (links with bold and italic) :
- Message 'version' => SXL_modules with { **id_module** / _family_ / version }
- Message 'componentList' => with { id_component / type="_family_/atypename" / descriptive_name }
- Messages Signals 'status'/'command'/'alarm' => with { CodeId="**id_module**/xyyyy" / cId (id_component adressing a component and its corresponding component type) }

**Comment by emiltin** [comment]
Hi @marcgarba, if I understand you correctly, the need you see is that a module might need to use component types from a diffrent module, which is not a parent? I believe this is similar to what i referred to above:

> If we need modules to be able to target component types defined in other modules, we could use a dependency mechanism. E.g. If module A depends on module B, it can target component types defined in B.

I'm not sure I completely understand your idea if "family". But I think we could handle the need without introducing a new concept of families. A module can just list other modules that it import types from. Examples

```
cits:
  define type S

tlc:
  define type X
  define type Y

tlc/advanced:
 inherit types from tlc    # becase tlc/advanced is a child of tlc, it inherits the types from tlc
 import types from cits    # import types from an unrelated module
 define type Z             # define a new type
```

The concept of inhering types from parent modules is just a conveniency, we could also choose to require all type imports to happen manually.

Defining or importanting a type just mean you can target it or refer to it when defining messages.

**Comment by marcgarba** [comment]
It seems to me that defining type families independently of signal modules is simpler and more flexible.
Each module could independently select the desired types of components

FamilyA : defines t1 , t2
FamilyB : defines t1 , t4
FamilyC : defines t9

Components in ModuleA : uses FamilyA/t1 & FamilyB/t1 & FamilyA/t4 ...
Components in ModuleB : uses FamilyA/t2 & FamilyC/t9 ...
(a component can use a type from any family)

**Comment by marcgarba** [comment]
Even if no more required for components types, this does not prevent the hierarchical naming of modules from being retained, if it's interesting. 
(allowing us to know that the existence of certain modules is linked to others)

However, it would be preferable to use another character (for example '>') other than the '/',  which should be reserved for separating the module part from the signal type/number.

<ins>Examples CodeId:</ins> 
"tlc **>** adanced **/** S5"
"detA **/** M4"
"detB **/** M4"
"detA **>** detAA **/** S8"
...

**Comment by emiltin** [comment]
> It seems to me that defining type families independently of signal modules is simpler and more flexible 

I don't really understand the benefit of a separate concept of 'familiy' for defining types. We would still have to manage dependencies because an sxl depends on families, buth with the added complexity of a new concept.
Beyond types, I think an SXL could also rely on messages defined in other SXLs. Not by using them directly, but by refering to them in the documentation, e.g. the docs for sxl B could explain that an advanced command expands on a basic command from sxl A, guide when to use which, etc. 

If you define a stand-along SXL it might be annoying to have to also define a familiy for the types?

In my mind we can use a model similar to how dependencies are handled in programming languages, where a library contains definitions of both types and functions.

> However, it would be preferable to use another character (for example '>') other than the '/', which should be reserved for separating the module part from the signal type/number.

Yes maybe. It might help avoid ambiguities in cases where messages are defined with the same name as SXLs. For example if:

- the sxl `tlc` defines the message called `advanced/m1`
- the sxl `tlc/advanced` defines the message `m1`

Then `tlc/advanced/m1` is ambiguous.

Whereas this is in principle not:

`<slx>:<subsxl>:<code>/<subcode>` 
`tlc:advanced/m1` 
`tlc:advanced:m1` 

However don't I think either of the above are very clear:

- with `tlc:advanced/m1 ` you might think that `tlc:advanced` is the sxl (actually the sxl is `tlc`)
- with `tlc:advanced:m1 ` it's not clear what part is the the sxl.

MQTT always uses slashes to separate levels, and let's you filter using wildcards. Since we plan to use MQTT in RSMP 4 I'm we might want to try to align the approaches.

One idea is to use extra empty level to separate sxls from codes:

`tlc//advanced/m1` 
`tlc/advanced//m1` 

Or add a colon the end of all sxls ids:

`tlc:/advanced/m1` 
`tlc:/advanced:/m1`

**Comment by marcgarba** [comment]
Returning to the link between component types and signal modules, here we have a concrete example specified : 
- a ‘traffic control’ group/module (with a french acronym!)
- a ‘public transport priority’ group/module

clearly separate and independent.
they just have one thing in common: addressing logical intersection components.

Perhaps we could simply have the component type prefix corresponding to the module name **by default**.
While still allowing the option to specify a different type prefix value for each module (or each component used in the module).... Flexibility for this type of case.

<ins>Example :</ins> 
Module "**TC**", with signals using components types "**TC**/aaaa", "**TC**/bbbb", "_INTERSEC_/inter"
Module "**PRIO**", with signals using components types "**PRIO**/aaaa", "**PRIO**/ccccc", "_INTERSEC_/inter"
(here with total flexibility: component per component)

As module names must be chosen carefully (no duplicates), we would have the same constraint for any additional component type prefixes (INTERSEC in the example).

**Comment by emiltin** [comment]
Yes we could take inspiration from how programming languages typically work with libraries/packages. A library has a name, but can define classes/types with arbitrary names. E.g. the package 'tlc' might define types 'tc', but also 'intersec', or any other type. The library name is not used when referring to types define in it.

This is flexible, but the downside is that there might be naming/type conflicts once you combine libraries, so you need to have a load/resolve step where you flag clashes, e.g. errors like "class A already defined".



> Module "TC", with signals using components types "TC/aaaa", "TC/bbbb", "INTERSEC/inter"
> Module "PRIO", with signals using components types "PRIO/aaaa", "PRIO/ccccc", "INTERSEC/inter"

Instead of allowing two sxls to define the same type I think a more conventional and cleaner model is to define it in one sxl, and let other sxls import it using a dependency mechanism. You could model it with dependencies either as:

```
traffic_light_control: define [tc, intersec]
public_priority: import [traffic_light_control],  define [prio]
```

```mermaid
graph RL
    subgraph traffic_light_control
        intersec[intersec]
        tlc[tlc]
        dl[tlc/dl]
    end

    subgraph public_priority
        prio[prio]
    end


   public_priority -->|depend| traffic_light_control
```

Or as:
```
intersection: define [intersec]
traffic_light_control: import [intersection], define [tlc]
public_priority: import [intersection], define [prio]
```

```mermaid
graph RL
    subgraph intersection
        intersec[intersec]
    end

    subgraph traffic_light_control
        tlc[tlc]
        dl[tlc/dl]
    end

    subgraph public_priority
        prio[prio]
    end

    traffic_light_control -->|depend| intersection
    public_priority -->|depend| intersection
```

I think it's ok to have an sxl that just defines types, for use as a depency for other modules.

I think it's good if we can follow established patterns for handling dependencies and types. An SXL seems similar to a library/package in a prgramming language. Is has a name, and can define types/functionality with arbitrary names. Of course they should be meaningful - a JSON library should not define an HTML module.


Types would not include the sxl name, e.g. just `tlc`, `tlc/dl`or `prio`. 
Command codes would not include the sxl name either. I guess they would be defined under a type? We would still need a way to delimit type from code, to avoid ambiguity, e.g. `tlc//advanced/m1`.

**Comment by emiltin** [comment]
But again, if we don't include the sxl name in types/codes, then you need context to understand references like `tc`. You need to know what slxs are used, to know where that particular type was defined. Or if you see `tc/m1`, you need to know the sxl to know what the code refers to and is supposed to do.
I guess this is similar to a situation where code might refer to a type or class, but you need to know which libraries where imported to understand what it really refers to. It would be impractical to have to include the package name in the code, i.e. you would not want to use "some_library/A" everywhere.

**Comment by marcgarba** [comment]
For signals codeID (it is different of "components types" problematic), in them we shoud simply always include the SXL module name as prefix.
(it is what we are doing here with the "group" of signals, and that seems to be more than enough)
This makes it easier to route signals to the right SXL, quickly However, the prefix must be fairly short, as it appears throughout RSMP messages...

For example : 
Module "**TC**" => signals CodeId : "**TC**/S1", "**TC**/M1", "**TC**/M2", ...
Module "**PRIO**" => signals CodeId : "**PRIO**/S1", "**PRIO**/S2", "**PRIO**/M1"...

**Comment by emiltin** [comment]
If we go for a library/package model:

## SXL names
An sxl has an name, e.g "traffic_light_controller" which must be globally unique. We maintain a registry of sxls.
An sxl is published using a semantic version, e.g. "2.5.3".

## IDs
Component types and codes for alarms, statuses and commands are identified using strings. Slashes can be used to create namespaces or otherwise organize ids, but no special format is mandated. Often an sxl will nest everything under a meaningful namespace, e..g 'tlc', but this is not required. 

Ids are must only be unique within the sxl, but you cannot use two sxls that define the same id.

This gives flexibility. For example, an sxl can add elements to an existing namespace or define things under several different namespaces.

Different sxls can also use the same top-levels. This makes it easy to create replacement sxls, but of course you cannot use both at the same time, as the define the same ids.

## Component Types
An sxl can define component types, identified by ids. No special form is mandated, and the sxl name is not used as a prefix. But often a top-level name is used which identifies the sxl or topic:

```
traffic_light_controller:
  types:
    tlc/tc
    tlc/dl
    tlc/sg
```

## Codes
Ids for types and codes for alarms, statuses and alarms all use the same namespace, so no name clashes are allowed. No special form is mandated, but ids are often nested under a common namespace.

```
traffic_light_controller:
  alarms:
    tlc/a1
    tlc/a2
  commands:
    tlc/plan/set (operates on components of type tlc/tc)
    tlc/dl/force  (operates on components of type tlc/dl)
  statuses:
    tlc/S0004
```

## Dependencies
An sxl can depend on other sxls. You can then  refer to types etc. define in the dependency.

```
public_priority:
  require:
    traffic_light_controller
  commands:
    tlc/prio/request (operates on components of type tlc/sg, defined in traffic_light_controller)
```

## Loading sxls
When a device or supervisor is configured, or a connection is established, the set of sxls is specified with versions. 
All ids in the used sxls can be collated into a single lookup table, since no id clashes are allowed.

```
sxls:
  system
  traffic_light_controller
  public_prioritty
```
=>
```
sys/version
sys/reset
tlc/a1
tlc/a
tlc/plan/set
tlc/dl/force
tlc/prio/request
tlc/S0004
```

## Messages
When sending a Command, Status or Alarm, you set the code, using the exact id defined in the relevant id e.g. 'tlc/plan/set'.

**Comment by emiltin** [comment]
> For signals codeID (it is different of "components types" problematic), in them we shoud simply always include the SXL module name as prefix. (it is what we are doing here with the "group" of signals, and that seems to be more than enough) This makes it easier to route signals to the right SXL, quickly However, the prefix must be fairly short, as it appears throughout RSMP messages...
> 
> For example : Module "**TC**" => signals CodeId : "**TC**/S1", "**TC**/M1", "**TC**/M2", ... Module "**PRIO**" => signals CodeId : "**PRIO**/S1", "**PRIO**/S2", "**PRIO**/M1"...

Yes. But with a package model, the SXL you have the freedom to use any prefix. you choose to use. This mean:
- you can split big sxl into multiple sxl, but keep using the same prefix
- you can extend another sxl
- you can create swap-in version replacements for an existing sxl

But as mentioned, it does require a load stage where you check that none of the loaded sxls define the same codes/types.
It also means that a prefix in itself does not indicate the sxl - you need to know which sxl defined the relevant code.

For your example:

Module "**traffic_light_controller**" => signals CodeId : "tc/s1", "tc/m1", "tc/m2", ... 
Module "**public_priority**" => signals CodeId : "prio/s1", "prio/s2", "prio/m1"...

The main difference is that the module name is independent from the prefixes you choose to use. This is just like a programming library; the name does not dictate what id defines, although you typically use constants that relate to the library name. In this example, the sxl traffic_light_controller uses the prefix 'tc', as an abbreviation of 'traffic controller'.

**Comment by emiltin** [comment]
More recent discsusions at #267,  #274, #275, #276, #277
